File: lib/python/openlinktoken/src/main/openlinktoken/tokens/inference_signature_provider.py
# SPDX-License-Identifier: MIT

from __future__ import annotations

from typing import Dict, List, Optional, Protocol, runtime_checkable

from openlinktoken.tokens.inference_batch_result import InferenceBatchResult


@runtime_checkable
class InferenceSignatureProvider(Protocol):
    """Protocol for inference-based token signature generation.

    Implementations are discovered at runtime via
    ``importlib.metadata.entry_points(group="openlinktoken.inference_providers")``.
    When no implementation is installed, inference-based tokens are silently
    disabled and only standard attribute-expression tokens (T1–T5) are generated.

    Each provider is responsible for one token identifier (e.g. ``"ML1"``) and
    reports it via :meth:`get_token_id`.
    """

    def get_token_id(self) -> str:
        """Return the token identifier this provider handles (e.g. ``"ML1"``)."""
        ...

    def is_enabled(self) -> bool:
        """Return whether this provider is currently enabled and configured."""
        ...

    def generate_signature(self, person_attributes: Dict[str, str]) -> Optional[str]:
        """Generate a single inference-based token signature.

        Args:
            person_attributes: Field-ID map for one record.

        Returns:
            Hex-encoded signature string, or ``None`` if the record is invalid.
        """
        ...

    def generate_batch(self, rows: List[Dict[str, str]]) -> InferenceBatchResult:
        """Generate inference-based signatures for a batch of records in one pass.

        Args:
            rows: List of field-ID maps, one per record.

        Returns:
            :class:`InferenceBatchResult` with signatures in input row order.
        """
        ...

File: lib/java/openlinktoken/src/main/java/org/openlinktoken/tokens/InferenceSignatureProvider.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokens;

import java.util.List;
import java.util.Map;

/**
 * SPI for inference-based token signature generation.
 *
 * <p>Implementations are discovered at runtime via {@link java.util.ServiceLoader}.
 * When no implementation is on the classpath, inference-based tokens are silently
 * disabled and only standard attribute-expression tokens (T1–T5) are generated.
 *
 * <p>Each provider is responsible for one token identifier (e.g. "ML1") and reports
 * it via {@link #getTokenId()}.
 */
public interface InferenceSignatureProvider {

    /**
     * Return the token identifier this provider handles (e.g. {@code "ML1"}).
     */
    String getTokenId();

    /**
     * Return whether this provider is currently enabled and configured.
     */
    boolean isEnabled();

    /**
     * Generate a single inference-based token signature from the given field-ID values.
     *
     * @param personAttributes field-ID map for one record
     * @return hex-encoded signature string, or {@code null} if the record is invalid
     */
    String generateSignature(Map<String, String> personAttributes);

    /**
     * Generate inference-based token signatures for a batch of records in a single
     * inference pass.
     *
     * @param rows list of field-ID maps, one per record
     * @return {@link InferenceBatchResult} with signatures in input row order
     */
    InferenceBatchResult generateBatch(List<Map<String, String>> rows);
}
